django_cloud_tasks/tasks/routine_task.py

```diff
@@ -0,0 +1,61 @@
+# pylint: disable=no-member
+import logging
+from typing import Dict
+from abc import abstractmethod
+from django.core.cache import cache
+from django_cloud_tasks import models, tasks
+
+logger = logging.getLogger()
+
+
+class RoutineTask(tasks.Task):
+    abstract = True
+
+    @abstractmethod
+    def run(self, **kwargs):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def revert(self, data: Dict, _meta: Dict, **kwargs):
+        routine = models.Routine.objects.get(pk=_meta.get("routine_id"))
+        routine.status = models.Routine.Statuses.REVERTED
+        routine.save()
+
+
+class PipelineRoutineTask(tasks.Task):
+    WAIT_FOR_LOCK = 5  # in seconds
+    LOCK_EXPIRATION = 60  # in seconds
+
+    def run(self, routine_id: int):
+        lock_key = f"lock-{self.__class__.__name__}-{routine_id}"
+        routine_lock = cache.lock(key=lock_key, timeout=self.LOCK_EXPIRATION, blocking_timeout=self.WAIT_FOR_LOCK)
+
+        with routine_lock:
+            self._run(routine_id=routine_id)
+
+    def _run(self, routine_id: int):
+        routine = models.Routine.objects.get(pk=routine_id)
+        if routine.status == models.Routine.Statuses.COMPLETED:
+            logger.info(f"Routine #{routine_id} is already completed")
+            return
+
+        if routine.retries_exhausted_limit and routine.attempt_count > routine.retries_exhausted_limit:
+            routine.pipeline.revert()
```

Comment:
A logger.info here would be nice too
```suggestion
            logger.info(f"Routine #{routine_id} has exhausted retries and is being reverted")
            routine.pipeline.revert()
```